### PR TITLE
Concretize factory ops

### LINKED ISSUE #720: tensor factories + broadcasting is failing
This issue is inspired by https://github.com/NVIDIA/Fuser/issues/718.  The following test is failing:

```C++
TEST_F(TensorFactoryTest, ARangeBroadcast) {
  auto fusion = std::make_unique<Fusion>();
  FusionGuard fg(fusion.get());

  Val* i = IrBuilder::create<Val>(DataType::Int);
  fusion->addInput(i);
  TensorView* tv0 = makeSymbolicTensor(1);
  fusion->addInput(tv0);

  Val* i_plus_one = add(i, fusion->oneVal(DataType::Int));
  auto tv1 = arange(i, i_plus_one);

  auto tv2 = add(tv1, tv0);
  fusion->addOutput(tv2);

  int64_t scalar = 15;

  const auto options =
      at::TensorOptions().dtype(at::kFloat).device(at::kCUDA, 0);
  auto input1 = at::randn({100}, options);

  FusionExecutorCache executor_cache(std::move(fusion));
  auto cg_outputs = executor_cache.runFusionWithInputs({scalar, input1});

  testValidate(
      executor_cache.fusion(),
      cg_outputs,
      {scalar, input1},
      {input1 + scalar},
      __LINE__,
      __FILE__);
}
```

Error message is:

```C++
[ RUN      ] TensorFactoryTest.ARangeBroadcast
unknown file: Failure
C++ exception with description "known_size == this_size INTERNAL ASSERT FAILED at "/home/gaoxiang/Fuser/csrc/expr_evaluator.cpp":206, please report a bug to PyTorch. Conflicting sizes: 100, 1
Exception raised from propagateBoundValuesThroughExactMaps at /home/gaoxiang/Fuser/csrc/expr_evaluator.cpp:206 (most recent call first):
frame #0: c10::Error::Error(c10::SourceLocation, std::string) + 0x57 (0x7f58803374c7 in /home/gaoxiang/.virtualenvs/nvfuser/lib/python3.11/site-packages/torch/lib/libc10.so)
frame #1: c10::detail::torchCheckFail(char const*, char const*, unsigned int, std::string const&) + 0x64 (0x7f58802f389b in /home/gaoxiang/.virtualenvs/nvfuser/lib/python3.11/site-packages/torch/lib/libc10.so)
frame #2: c10::detail::torchInternalAssertFail(char const*, char const*, unsigned int, char const*, std::string const&) + 0x3f (0x7f588033511f in /home/gaoxiang/.virtualenvs/nvfuser/lib/python3.11/site-packages/torch/lib/libc10.so)
frame #3: nvfuser::ExpressionEvaluator::propagateBoundValuesThroughExactMaps(nvfuser::Fusion*) + 0x70f (0x7f587fc368df in /home/gaoxiang/Fuser/build/libnvfuser_codegen.so)
frame #4: nvfuser::DynamicTransformConcretizationInfo::DynamicTransformConcretizationInfo(nvfuser::DynamicTransformInitialInfo const*, nvfuser::ExpressionEvaluator*) + 0xa8 (0x7f587fc2f8c8 in /home/gaoxiang/Fuser/build/libnvfuser_codegen.so)
frame #5: nvfuser::FusionExecutorCache::getKernelRuntimeFor(nvfuser::KernelArgumentHolder const&, std::optional<nvfuser::PrimDataType>) + 0x161 (0x7f587fd91f31 in /home/gaoxiang/Fuser/build/libnvfuser_codegen.so)
frame #6: nvfuser::FusionExecutorCache::runFusionWithInputs(c10::ArrayRef<c10::IValue> const&, std::optional<nvfuser::PrimDataType>, std::optional<signed char>) + 0x1fd (0x7f587fd931ad in /home/gaoxiang/Fuser/build/libnvfuser_codegen.so)
frame #7: <unknown function> + 0x4b9f7d (0x55e4af7d0f7d in ./build/nvfuser_tests)
frame #8: <unknown function> + 0x695c47 (0x55e4af9acc47 in ./build/nvfuser_tests)
frame #9: <unknown function> + 0x6884fd (0x55e4af99f4fd in ./build/nvfuser_tests)
frame #10: <unknown function> + 0x6886d5 (0x55e4af99f6d5 in ./build/nvfuser_tests)
frame #11: <unknown function> + 0x6888b0 (0x55e4af99f8b0 in ./build/nvfuser_tests)
frame #12: <unknown function> + 0x68c08d (0x55e4af9a308d in ./build/nvfuser_tests)
frame #13: <unknown function> + 0x6961d7 (0x55e4af9ad1d7 in ./build/nvfuser_tests)
frame #14: <unknown function> + 0x688ae2 (0x55e4af99fae2 in ./build/nvfuser_tests)
frame #15: <unknown function> + 0x193279 (0x55e4af4aa279 in ./build/nvfuser_tests)
frame #16: <unknown function> + 0x23850 (0x7f5879e39850 in /usr/lib/libc.so.6)
frame #17: __libc_start_main + 0x8a (0x7f5879e3990a in /usr/lib/libc.so.6)
frame #18: <unknown function> + 0x1cb5a5 (0x55e4af4e25a5 in ./build/nvfuser_tests)
" thrown in the test body.
[  FAILED  ] TensorFactoryTest.ARangeBroadcast (62 ms)
```

in the above test, we created a tensor in the fusion whose size is `i + 1 - i`, which is just `1`, and broadcast this tensor and add it to another tensor. However, our system is unable to infer the size as one because it is dynamic. Fortunately, for the case of this test, we can just `simplifyExpr` the extent to get a `1` and mark that ID as broadcasting, however, `simplifyExpr` does not solve all problems. For example in this following test:

```C++
TEST_F(TensorFactoryTest, ARangeBroadcast2) {
  auto fusion = std::make_unique<Fusion>();
  FusionGuard fg(fusion.get());

  Val* i = IrBuilder::create<Val>(DataType::Int);
  fusion->addInput(i);
  TensorView* tv0 = makeSymbolicTensor(1);
  fusion->addInput(tv0);

  auto tv1 = arange(i);

  auto tv2 = add(tv1, tv0);
  fusion->addOutput(tv2);

  const auto options =
      at::TensorOptions().dtype(at::kFloat).device(at::kCUDA, 0);
  auto input1 = at::randn({100}, options);

  FusionExecutorCache executor_cache(std::move(fusion));
  auto cg_outputs = executor_cache.runFusionWithInputs({1, input1});

  testValidate(
      executor_cache.fusion(),
      cg_outputs,
      {1, input1},
      {input1},
      __LINE__,
      __FILE__);
}
```
The size depend on the input, so there is no way to know if the `add` op in the fusion is a broadcasting add or not, because the actual computation depends on the input.

### PR DESCRIPTION
**Pull request by jacobhinkle**
Factory ops `full`, `iota`, `uniform`, `normal`, `rand`, and `randn` take scalar inputs that determine their output shape. At definition, these scalars might be undetermined, meaning the output shape might contain 1 values. This PR marks non-constant extent axes as `Symbolic` at definition then concretizes their `IterType`s later once we have inputs.

Fixes #720.

### CODE REVIEW

**Comment by jacobhinkle** [comment]
!build --diff

**Comment by jacobhinkle** [comment]
!build --diff

**Comment by jacobhinkle** [comment]
!build --diff

**Comment by jacobhinkle** [comment]
Marking as draft until I fix the test failures. Some of these are just tests that (wrongly) assume the fusion should not be dynamic, but others are real.

**Comment by jacobhinkle** [comment]
!build --diff

**Comment by jacobhinkle** [comment]
!build --diff

File: csrc/preseg_passes/remove_empty.cpp [hunk 1]
```diff
@@ -104,7 +104,7 @@ class EmptyTensorRemover : public DeadCodeRemover {
           tv->toString());
       auto shape = noReductionShape(tv);
       auto dtype = tv->getDataType().value();
-      auto new_tv = zeros(shape, dtype);
+      auto new_tv = zeros(shape, dtype, /*maybe_symbolic=*/false);
```
Comment (jacobhinkle) [inline csrc/preseg_passes/remove_empty.cpp:107]: Without the `maybe_symbolic=false` arguments, this pass would re-introduce symbolic IDs by calling factory methods `full` and `zeros`. That caused numerous test failures, such as in `StandaloneEye`. This argument lets us default to using `Iteration` for symbolic axes in those factory ops.

**Comment by jacobhinkle** [comment]
!build --diff

**Comment by jacobhinkle** [comment]
!build --diff

**Comment by jacobhinkle** [comment]
!build --diff

**Comment by jacobhinkle** [comment]
!build --diff

**Comment by jacobhinkle** [comment]
!build --diff

**Comment by xwang233** [comment]
!build

**Comment by jacobhinkle** [comment]
!build --diff

**Comment by jacobhinkle** [comment]
CI showed a broken test which is fixed. codediff showed there is an effect on segmentation in the standalone tests e.g. `StandaloneFull`. I think this is because we are now rightly setting some axes as `Broadcast` when their size is 1, but I want to understand why this causes more segments. Other code diffs look like changes in variable naming but I will verify that.

**Comment by jacobhinkle** [comment]
> why this causes more segments.

I was mistaken. This doesn't cause more segments. It _does_ cause more kernels to be generated, but that is because whereas previously we would have two concretized kernels (one empty and one Iteration), we now have 3 concretizations since size of 1 now causes `Broadcast` axes. I think this is just working as intended.

**Comment by jacobhinkle** [comment]
!build --diff-bench

**Comment by jacobhinkle** [comment]
Test failure on V100 is an OOM at `IndexingOpTest.TorchGatherAddMul_CUDA`. That is unrelated to this PR.

**Comment by jacobhinkle** [comment]
!build --diff

**Comment by jacobhinkle** [comment]
Benchmark failures are unrelated and the code-diff looks reasonable. I believe this is ready, but I don't think it is urgent.

**Review by naoyam — APPROVED** [review]
LGTM. Thanks for adding this support.

**Pull request merged**